File: kotlin/app/src/test/java/com/vitalcheck/app/domain/entity/SymptomEntryTest.kt
package com.vitalcheck.app.domain.entity

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Testes unitários para a entidade SymptomEntry.
 * Validação de invariantes, factory e limites.
 */
class SymptomEntryTest {

    @Test
    fun `deve criar SymptomEntry com valores validos`() {
        val entry = SymptomEntry(id = "abc", description = "Dor de cabeça")
        assertEquals("abc", entry.id)
        assertEquals("Dor de cabeça", entry.description)
    }

    @Test(expected = IllegalArgumentException::class)
    fun `deve rejeitar descricao vazia`() {
        SymptomEntry(id = "abc", description = "")
    }

    @Test(expected = IllegalArgumentException::class)
    fun `deve rejeitar descricao apenas com espacos`() {
        SymptomEntry(id = "abc", description = "   ")
    }

    @Test(expected = IllegalArgumentException::class)
    fun `deve rejeitar descricao com mais de 500 caracteres`() {
        val longDescription = "a".repeat(501)
        SymptomEntry(id = "abc", description = longDescription)
    }

    @Test
    fun `deve aceitar descricao com exatamente 500 caracteres`() {
        val description = "a".repeat(500)
        val entry = SymptomEntry(id = "abc", description = description)
        assertEquals(500, entry.description.length)
    }

    @Test(expected = IllegalArgumentException::class)
    fun `deve rejeitar ID vazio`() {
        SymptomEntry(id = "", description = "Dor de cabeça")
    }

    @Test(expected = IllegalArgumentException::class)
    fun `deve rejeitar ID apenas com espacos`() {
        SymptomEntry(id = "   ", description = "Dor de cabeça")
    }

    @Test
    fun `factory create deve fazer trim na descricao`() {
        val entry = SymptomEntry.create(description = "  Dor de cabeça  ")
        assertEquals("Dor de cabeça", entry.description)
    }

    @Test
    fun `factory create deve gerar ID automaticamente`() {
        val entry = SymptomEntry.create(description = "Tontura")
        assertTrue(entry.id.isNotBlank())
    }

    @Test
    fun `factory create deve gerar IDs unicos`() {
        val entry1 = SymptomEntry.create(description = "Tontura")
        val entry2 = SymptomEntry.create(description = "Febre")
        assertTrue(entry1.id != entry2.id)
    }

    @Test
    fun `factory create deve gerar timestamp automaticamente`() {
        val before = System.currentTimeMillis()
        val entry = SymptomEntry.create(description = "Dor")
        val after = System.currentTimeMillis()
        assertTrue(entry.timestamp in before..after)
    }
}

File: src/__tests__/domain/entities/SymptomEntry.test.ts
import { createSymptomEntry } from '../../../domain/entities/SymptomEntry';

describe('SymptomEntry Entity', () => {
  describe('createSymptomEntry', () => {
    it('deve criar um SymptomEntry válido', () => {
      const timestamp = new Date('2026-02-10T15:00:00Z');
      const entry = createSymptomEntry('abc-123', 'Dor de cabeça leve', timestamp);

      expect(entry.id).toBe('abc-123');
      expect(entry.description).toBe('Dor de cabeça leve');
      expect(entry.timestamp).toBe(timestamp);
    });

    it('deve fazer trim na descrição', () => {
      const entry = createSymptomEntry('id-1', '  Dor de cabeça  ');
      expect(entry.description).toBe('Dor de cabeça');
    });

    it('deve criar objeto imutável', () => {
      const entry = createSymptomEntry('id-1', 'Tontura');

      expect(Object.isFrozen(entry)).toBe(true);
      expect(() => {
        (entry as any).description = 'Alterado';
      }).toThrow();
    });

    it('deve usar timestamp atual quando não fornecido', () => {
      const before = new Date();
      const entry = createSymptomEntry('id-1', 'Náusea');
      const after = new Date();

      expect(entry.timestamp.getTime()).toBeGreaterThanOrEqual(before.getTime());
      expect(entry.timestamp.getTime()).toBeLessThanOrEqual(after.getTime());
    });

    it('deve rejeitar descrição vazia', () => {
      expect(() => createSymptomEntry('id-1', '')).toThrow(
        'A descrição do sintoma não pode ser vazia'
      );
    });

    it('deve rejeitar descrição apenas com espaços', () => {
      expect(() => createSymptomEntry('id-1', '   ')).toThrow(
        'A descrição do sintoma não pode ser vazia'
      );
    });

    it('deve rejeitar descrição com mais de 500 caracteres', () => {
      const longDescription = 'A'.repeat(501);
      expect(() => createSymptomEntry('id-1', longDescription)).toThrow(
        'não pode exceder 500 caracteres'
      );
    });

    it('deve aceitar descrição com exatamente 500 caracteres', () => {
      const maxDescription = 'A'.repeat(500);
      const entry = createSymptomEntry('id-1', maxDescription);
      expect(entry.description.length).toBe(500);
    });

    it('deve rejeitar ID vazio', () => {
      expect(() => createSymptomEntry('', 'Dor')).toThrow(
        'O ID do sintoma não pode ser vazio'
      );
    });
  });
});

